rllib/agents/alpha_star/league_builder.py

```diff
@@ -0,0 +1,354 @@
+from abc import ABCMeta
+import logging
+import numpy as np
+import re
+
+from ray.rllib.agents.trainer import Trainer
+from ray.rllib.examples.policy.random_policy import RandomPolicy
+from ray.rllib.policy.policy import PolicySpec
+from ray.rllib.utils.annotations import ExperimentalAPI, override
+from ray.rllib.utils.numpy import softmax
+from ray.rllib.utils.typing import TrainerConfigDict, ResultDict
+
+logger = logging.getLogger(__name__)
+
+
+@ExperimentalAPI
+class LeagueBuilder(metaclass=ABCMeta):
+    def __init__(self, trainer: Trainer, trainer_config: TrainerConfigDict):
+        """Initializes a LeagueBuilder instance.
+
+        Args:
+            trainer: The Trainer object by which this league builder is used.
+                Trainer calls `build_league()` after each training step.
+            trainer_config: The (not yet validated) config dict to be
+                used on the Trainer. Child classes of `LeagueBuilder`
+                should preprocess this to add e.g. multiagent settings
+                to this config.
+        """
+        self.trainer = trainer
+        self.config = trainer_config
+
+    def build_league(self, result: ResultDict) -> None:
+        """Method containing league-building logic. Called after train step.
+
+        Args:
+            result: The most recent result dict with all necessary stats in
+                it (e.g. episode rewards) to perform league building
+                operations.
+        """
+        raise NotImplementedError
+
+
+@ExperimentalAPI
+class NoLeagueBuilder(LeagueBuilder):
+    """A LeagueBuilder that does nothing.
+
+    Useful for simple non-league-building self-play setups.
```

Comment:
I'm not sure if I understand -- can you give an example?

Reply:
Any multi-agent environment whose solution does not necessarily require league-building.
See e.g.: `tuned_examples/alpha_star/multi-agent-cart-pole-alpha-star.yaml`

Reply:
I should remove the "self-play". AlphaStar works on non self-play problems as well.

Reply:
done